src/Fcf/Data/Symbol.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE DataKinds              #-}
+{-# LANGUAGE PolyKinds              #-}
+{-# LANGUAGE TypeFamilies           #-}
+{-# LANGUAGE TypeInType             #-}
+{-# LANGUAGE TypeOperators          #-}
+{-# LANGUAGE UndecidableInstances   #-}
+{-# OPTIONS_GHC -Wall                       #-}
+{-# OPTIONS_GHC -Werror=incomplete-patterns #-}
+
+{-|
+Module      : Fcf.Data.Symbol
+Description : Type level symbols
+
+= Fcf.Data.Symbol
+
+Type-level symbols and functions for them.
+
+Note that the operators from this module conflict with "GHC.TypeLits".
+
+-}
+
+--------------------------------------------------------------------------------
+
+module Fcf.Data.Symbol
+    ( -- * Reexported type
+      -- | From "GHC.TypeLits".
+
+      Symbol
+
+     -- * Comparison functions
+
+    , SymbolOrd
+    , type (<=)
+    , type (>=)
+    , type (<)
+    , type (>)
+    , type (==)
+    )
+  where
+
+--------------------------------------------------------------------------------
+        
+import           GHC.TypeLits (Symbol)
+import qualified GHC.TypeLits as TL
+
+import           Fcf.Core (Eval, Exp)
+import           Fcf.Data.Bool (type (||))
+import           Fcf.Utils (TyEq)
+
+--------------------------------------------------------------------------------
+
+
+-- | SymbolOrd - compare two symbols and give type-level Ordering 
+-- ( $ 'LT $, $ 'EQ $ or $ 'GT $ ).
+--
+-- === __Example__
+-- 
+-- >>> :kind! Eval (SymbolOrd "a" "b")
+-- Eval (SymbolOrd "a" "b") :: Ordering
+-- = 'LT
+data SymbolOrd :: Symbol -> Symbol -> Exp Ordering
+type instance Eval (SymbolOrd a b) = TL.CmpSymbol a b
+
+-- | Less-than-or-equal comparison for symbols.
+-- 
+-- === __Example__
+-- 
+-- >>> :kind! Eval ("b" <= "a")
+-- Eval ("b" <= "a") :: Bool
+-- = 'False
+--
+data (<=) :: Symbol -> Symbol -> Exp Bool
+type instance Eval ((<=) a b) =
+  Eval (Eval (TyEq (TL.CmpSymbol a b) 'LT) || Eval (TyEq (TL.CmpSymbol a b) 'EQ))
```

Comment:
Make a specialized type family `IsLE :: Ordering -> Bool` instead of using `(||)`, avoiding some duplication.

Reply:
done